src/blockmap.jl

```diff
@@ -0,0 +1,477 @@
+        new{T, R, BS}(maps, block_sizes)
+    end
+end
+
+struct VBlockMap{T,As<:Tuple{Vararg{LinearMap}},BS<:AbstractBlockSizes} <: AbstractBlockMap{T,As,BS}
+    maps::As
+    block_sizes::BS
+    global function _VBlockMap(maps::R, block_sizes::BS) where {T, R<:Tuple{Vararg{LinearMap{T}}}, BS<:AbstractBlockSizes}
+        new{T, R, BS}(maps, block_sizes)
+    end
+end
+
+# BlockMap(maps::Tuple{Vararg{LinearMap}}, rows::Integer) = isone(rows) ? _HBlockMap(maps, sizes_from_maps(maps, rows))
+HBlockMap(maps::Tuple{Vararg{LinearMap{T}}}) where {T} = _HBlockMap(maps, sizes_from_maps(maps, 1))
+VBlockMap(maps::Tuple{Vararg{LinearMap{T}}}) where {T} = _VBlockMap(maps, sizes_from_maps(maps, length(maps)))
+
+function sizes_from_maps(maps::Tuple{Vararg{LinearMap}}, rows::Integer)
+	N = length(maps)
+	if N == 0
+        return zeros.(Int, size(maps))
+    end
+	cols, rem = divrem(N, rows)
+    if rem != 0
+        error("Cannot construct a BlockMap with $rows block rows from $N linear maps")
+    end
+    fullsizes::Array{Tuple{Int,Int}} = reshape([map(size, maps)...,], rows, cols)
+    block_sizes = ntuple(2) do i
+        [s[i] for s in view(fullsizes, ntuple(j -> j == i ? (:) : 1, 2)...)]
+    end
+    checksizes(fullsizes, block_sizes)
+    return BlockSizes(block_sizes...)
+end
+
+getsizes(block_sizes, block_index) = getindex.(block_sizes, block_index)
+
+function checksizes(fullsizes::Matrix{Tuple{Int,Int}}, block_sizes::Tuple{Vector{Int},Vector{Int}})
+    Base.@nloops 2 i fullsizes begin
+        block_index = Base.@ntuple 2 i
+        if fullsizes[block_index...] != getsizes(block_sizes, block_index)
+			error("size(blocks[", strip(repr(block_index), ['(', ')']),
+                      "]) (= ", fullsizes[block_index...],
+                      ") is incompatible with expected size: ",
+                      getsizes(block_sizes, block_index))
+        end
+    end
+    return fullsizes
+end
+
+@inline Base.size(A::AbstractBlockMap) = size(blocksizes(A))
+
+"""
+    blocksizes(A)
+
+Returns a subtype of `AbstractBlockSizes` that contains information about the
+block sizes of `A`.
+"""
+@inline blocksizes(A::AbstractBlockMap) = A.block_sizes
+
+@inline function blockcheckbounds(A::AbstractBlockMap{T}, i::Integer, j::Integer) where {T}
+    n = nblocks(A)
+	if i <= 0 || i > n[1]
+		return throw(BoundsError(A, (i, j)))
+	end
+	if j <= 0 || j > n[2]
+		return throw(BoundsError(A, (i, j)))
+	end
+    return nothing
+end
+
+@inline nblocks(A::AbstractBlockMap) = nblocks(blocksizes(A))
+
+nblocks(A::AbstractBlockMap, i::Integer) = nblocks(A)[i]
+
+nblocks(A::AbstractBlockMap, i::Integer, j::Integer) = nblocks(blocksizes(A), i, j)
+
+@inline function getblock(A::AbstractBlockMap, i::Integer, j::Integer)
+    @boundscheck blockcheckbounds(A, i, j)
+	m, n = nblocks(A)
+    A.maps[(j-1)*m + i]
+end
+
+############
+# hcat
+############
+
+function Base.hcat(As::Union{LinearMap,UniformScaling}...)
```

Comment:
I happily accept approaches to how to avoid this. One simple way to avoid this is to get rid of convenience, and force the user to provide `LinearMaps`, but I don't think we want to go that way. We would need to export the `UniformScalingMap` constructor.